lib/hcrypto/evp-pkcs11.c

```diff
@@ -0,0 +1,744 @@
+# define PKCS11_MODULE_PATH "/usr/lib/64/libpkcs11.so"
+# else
+# define PKCS11_MODULE_PATH "/usr/lib/libpkcs11.so"
+# endif
+#endif
+
+static CK_FUNCTION_LIST_PTR p11_module;
+
+static int
+p11_cleanup(EVP_CIPHER_CTX *ctx);
+
+struct pkcs11_cipher_ctx {
+    CK_SESSION_HANDLE hSession;
+    CK_OBJECT_HANDLE hSecret;
+};
+
+struct pkcs11_md_ctx {
+    CK_SESSION_HANDLE hSession;
+};
+
+static CK_RV
+p11_module_init(void)
+{
+    CK_RV rv = CKR_OK;
+
+    if (p11_module == NULL) {
+        CK_RV (*C_GetFunctionList_fn)(CK_FUNCTION_LIST_PTR_PTR);
+        CK_FUNCTION_LIST_PTR module;
+        void *handle = NULL;
+
+        if (!issuid()) {
+            char *pkcs11ModulePath = getenv("PKCS11_MODULE_PATH");
+            if (pkcs11ModulePath != NULL) {
+                handle = dlopen(pkcs11ModulePath, RTLD_LAZY | RTLD_LOCAL | RTLD_GROUP);
+                if (handle == NULL)
+                    fprintf(stderr, "p11_module_init(%s): %s\n", pkcs11ModulePath, dlerror());
+            }
+        }
+#ifdef PKCS11_MODULE_PATH
+        if (handle == NULL) {
+            handle = dlopen(PKCS11_MODULE_PATH, RTLD_LAZY | RTLD_LOCAL | RTLD_GROUP);
+            if (handle == NULL)
+                fprintf(stderr, "p11_module_init(%s): %s\n", PKCS11_MODULE_PATH, dlerror());
+        }
+#endif
+        if (handle == NULL) {
+            rv = CKR_LIBRARY_LOAD_FAILED;
+            goto cleanup;
+        }
+
+        C_GetFunctionList_fn = (CK_RV (*)(CK_FUNCTION_LIST_PTR_PTR))
+                               dlsym(handle, "C_GetFunctionList");
+        if (C_GetFunctionList_fn == NULL) {
+            rv = CKR_LIBRARY_LOAD_FAILED;
+            goto cleanup;
+        }
+
+        rv = C_GetFunctionList_fn(&module);
+        if (rv != CKR_OK)
+            goto cleanup;
+
+        rv = module->C_Initialize(NULL);
+        if (rv == CKR_CRYPTOKI_ALREADY_INITIALIZED)
+            rv = CKR_OK;
+        if (rv == CKR_OK)
+            (void)heim_base_exchange_pointer(&p11_module, module);
+
+cleanup:
+        if (handle != NULL && p11_module == NULL)
+            dlclose(handle);
+        /* else leak handle */
+    }
+
+    return rv;
+}
+
+static CK_RV
+p11_session_init(CK_MECHANISM_TYPE mechanismType, CK_SESSION_HANDLE_PTR phSession)
+{
+    CK_RV rv;
+    CK_ULONG i, ulSlotCount = 0;
+    CK_SLOT_ID_PTR pSlotList = NULL;
+    CK_MECHANISM_INFO info;
+
+    if (phSession != NULL)
+        *phSession = CK_INVALID_HANDLE;
+
+    rv = p11_module_init();
+    if (rv != CKR_OK)
+        goto cleanup;
+
+    assert(p11_module != NULL);
+
+    rv = p11_module->C_GetSlotList(CK_FALSE, NULL, &ulSlotCount);
+    if (rv != CKR_OK)
+        goto cleanup;
+
+    pSlotList = (CK_SLOT_ID_PTR)calloc(ulSlotCount, sizeof(CK_SLOT_ID));
+    if (pSlotList == NULL) {
+        rv = CKR_HOST_MEMORY;
+        goto cleanup;
+    }
+
+    rv = p11_module->C_GetSlotList(CK_FALSE, pSlotList, &ulSlotCount);
+    if (rv != CKR_OK)
+        goto cleanup;
+
+    for (i = 0; i < ulSlotCount; i++) {
```

Comment:
Well, you will get the first slot that works even if there's no meta-slot provider but I agree it may not get you the fastest or more desired provider.

Sent from my iPhone

> On 29 Nov 2015, at 11:22, Nico Williams <notifications@github.com> wrote:
> 
> In lib/hcrypto/evp-pkcs11.c:
> 
> > +
> > +    rv = p11_module->C_GetSlotList(CK_FALSE, NULL, &ulSlotCount);
> > +    if (rv != CKR_OK)
> > +        goto cleanup;
> > +
> > +    pSlotList = (CK_SLOT_ID_PTR)calloc(ulSlotCount, sizeof(CK_SLOT_ID));
> > +    if (pSlotList == NULL) {
> > +        rv = CKR_HOST_MEMORY;
> > +        goto cleanup;
> > +    }
> > +
> > +    rv = p11_module->C_GetSlotList(CK_FALSE, pSlotList, &ulSlotCount);
> > +    if (rv != CKR_OK)
> > +        goto cleanup;
> > +
> > +    for (i = 0; i < ulSlotCount; i++) {
> I realize that this works on Solaris, and probably elsewhere by accident. If we're going to allow specifying a provider by env var and/or krb5.conf, we might as well also allow a slot to be specified. But don't implement this. Let's leave a cookie crumb explaining that this is a bit of a hack and that it's not clear what's the best thing to do.
> 
> There's an RFC for PKCS#11 URIs, RFC7512, that ideally we could use for all of this, perhaps with even a URI per-mechanism.
> 
> —
> Reply to this email directly or view it on GitHub.
>